From 697f2dc065b0a0dd3ce808f96f508872722dfa92 Mon Sep 17 00:00:00 2001
From: Yi Zhao <yi.zhao@windriver.com>
Date: Wed, 19 Jun 2019 14:20:31 +0800
Subject: [PATCH] Revert "Remove unused translate permission in context
 userspace class."

This reverts commit 65da822c1b5c70bd1ff7eca645f5f9fd74fa949e.

The refpolicy commit 65da822c1b5c70bd1ff7eca645f5f9fd74fa949e introduced
an issue that mcstransd can not translate SELinux MCS/MLS sensitivity
labels:

root@qemux86-64:~# id -Z
root:sysadm_r:sysadm_t:s0-s15:c0.c1023

It should be:
root@qemux86-64:~# id -Z
root:sysadm_r:sysadm_t:SystemLow-SystemHigh

Revert this commit to fix the issue.

Fix avc denials:

type=AVC msg=audit(1560928397.479:270): avc:  denied  { write } for
pid=309 comm="login" name=".setrans-unix" dev="tmpfs" ino=14099
scontext=system_u:system_r:local_login_t:s0-s15:c0.c1023
tcontext=system_u:object_r:setrans_var_run_t:s15:c0.c1023
tclass=sock_file permissive=0
type=AVC msg=audit(1560928399.141:272): avc:  denied  { connectto } for
pid=2416 comm="id" path="/run/setrans/.setrans-unix"
scontext=root:sysadm_r:sysadm_t:s0-s15:c0.c1023
tcontext=system_u:system_r:setrans_t:s15:c0.c1023
tclass=unix_stream_socket permissive=0

Upstream-Status: Pending

Signed-off-by: Yi Zhao <yi.zhao@windriver.com>
---
 policy/flask/access_vectors      |  2 +-
 policy/mls                       |  3 +++
 policy/modules/kernel/domain.te  |  4 ++++
 policy/modules/kernel/mls.if     |  8 ++++++--
 policy/modules/kernel/mls.te     |  2 ++
 policy/modules/system/setrans.if | 12 ++++++++++--
 6 files changed, 26 insertions(+), 5 deletions(-)

diff --git a/policy/flask/access_vectors b/policy/flask/access_vectors
index 40d1c24..14d1fa9 100644
--- a/policy/flask/access_vectors
+++ b/policy/flask/access_vectors
@@ -765,7 +765,7 @@ class key
 
 class context
 {
-	unused_perm
+	translate
 	contains
 }
 
diff --git a/policy/mls b/policy/mls
index 484e3ca..eeca15a 100644
--- a/policy/mls
+++ b/policy/mls
@@ -764,6 +764,9 @@ mlsconstrain association { polmatch }
 # MLS policy for the context class
 #
 
+mlsconstrain context translate
+	(( h1 dom h2 ) or ( t1 == mlstranslate ));
+
 mlsconstrain context contains
 	(( h1 dom h2 ) and ( l1 domby l2));
 
diff --git a/policy/modules/kernel/domain.te b/policy/modules/kernel/domain.te
index 15df92b..097b254 100644
--- a/policy/modules/kernel/domain.te
+++ b/policy/modules/kernel/domain.te
@@ -152,6 +152,10 @@ optional_policy(`
 	libs_use_shared_libs(domain)
 ')
 
+optional_policy(`
+	setrans_translate_context(domain)
+')
+
 # xdm passes an open file descriptor to xsession-errors.log which is then audited by all confined domains.
 optional_policy(`
 	xserver_dontaudit_use_xdm_fds(domain)
diff --git a/policy/modules/kernel/mls.if b/policy/modules/kernel/mls.if
index c11c7b9..2e2bebc 100644
--- a/policy/modules/kernel/mls.if
+++ b/policy/modules/kernel/mls.if
@@ -849,7 +849,7 @@ interface(`mls_fd_share_all_levels',`
 ########################################
 ## <summary>
 ##	Make specified domain MLS trusted
-##	for translating contexts at all levels.  (Deprecated)
+##	for translating contexts at all levels.
 ## </summary>
 ## <param name="domain">
 ##	<summary>
@@ -859,7 +859,11 @@ interface(`mls_fd_share_all_levels',`
 ## <rolecap/>
 #
 interface(`mls_context_translate_all_levels',`
-	refpolicywarn(`$0($*) has been deprecated')
+	gen_require(`
+		attribute mlstranslate;
+	')
+
+	typeattribute $1 mlstranslate;
 ')
 
 ########################################
diff --git a/policy/modules/kernel/mls.te b/policy/modules/kernel/mls.te
index e4bc214..868380e 100644
--- a/policy/modules/kernel/mls.te
+++ b/policy/modules/kernel/mls.te
@@ -69,5 +69,7 @@ attribute mlsrangetrans;
 attribute mlsfduse;
 attribute mlsfdshare;
 
+attribute mlstranslate;
+
 attribute mlsdbusrecv;
 attribute mlsdbussend;
diff --git a/policy/modules/system/setrans.if b/policy/modules/system/setrans.if
index 03afaa9..9478dd9 100644
--- a/policy/modules/system/setrans.if
+++ b/policy/modules/system/setrans.if
@@ -21,7 +21,7 @@ interface(`setrans_initrc_domtrans',`
 
 #######################################
 ## <summary>
-##	Allow a domain to translate contexts.  (Deprecated)
+##	Allow a domain to translate contexts.
 ## </summary>
 ## <param name="domain">
 ##	<summary>
@@ -30,7 +30,15 @@ interface(`setrans_initrc_domtrans',`
 ## </param>
 #
 interface(`setrans_translate_context',`
-	refpolicywarn(`$0($*) has been deprecated')
+	gen_require(`
+		type setrans_t, setrans_var_run_t;
+		class context translate;
+	')
+
+	allow $1 self:unix_stream_socket create_stream_socket_perms;
+	allow $1 setrans_t:context translate;
+	stream_connect_pattern($1, setrans_var_run_t, setrans_var_run_t, setrans_t)
+	files_list_pids($1)
 ')
 
 ######################################
-- 
2.7.4

